crypto/bp/bp_g2_mult.c

```diff
@@ -0,0 +1,701 @@
+{
+    const G2_ELEM *points[1];
+    const BIGNUM *scalars[1];
+
+    points[0] = point;
+    scalars[0] = p_scalar;
+    return G2_ELEMs_mul(group, r, g_scalar,
+                        (point != NULL
+                         && p_scalar != NULL), points, scalars, ctx);
+
+}
+
+/*-
+ * Compute
+ *      \sum scalars[i]*points[i],
+ * also including
+ *      scalar*generator
+ * in the addition if scalar != NULL
+ */
+int G2_ELEMs_mul(const BP_GROUP *group, G2_ELEM *r, const BIGNUM *scalar,
+                 size_t num, const G2_ELEM *points[],
+                 const BIGNUM *scalars[], BN_CTX *ctx)
+{
+    BN_CTX *new_ctx = NULL;
+    const G2_ELEM *generator = NULL;
+    G2_ELEM *tmp = NULL;
+    size_t totalnum;
+    size_t blocksize = 0, numblocks = 0; /* for wNAF splitting */
+    size_t pre_points_per_block = 0;
+    size_t i, j;
+    int k;
+    int r_is_inverted = 0;
+    int r_is_at_infinity = 1;
+    size_t *wsize = NULL;       /* individual window sizes */
+    signed char **wNAF = NULL;  /* individual wNAFs */
+    size_t *wNAF_len = NULL;
+    size_t max_len = 0;
+    size_t num_val;
+    G2_ELEM **val = NULL;       /* precomputation */
+    G2_ELEM **v;
+    G2_ELEM ***val_sub = NULL;  /* pointers to sub-arrays of 'val' or
+                                 * 'pre_comp->points' */
+    const G2_PRE_COMP *pre_comp = NULL;
+    int num_scalar = 0;         /* flag: will be set to 1 if 'scalar' must be
+                                 * treated like other scalars, i.e.
+                                 * precomputation is not available */
+    int ret = 0;
+
+    if (scalar == NULL && num == 0)
+        return G2_ELEM_set_to_infinity(group, r);
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_new()) == NULL)
+        return 0;
+
+    if (scalar != NULL) {
+        generator = group->gen2;
+        if (generator == NULL) {
+            ECerr(EC_F_EC_WNAF_MUL, EC_R_UNDEFINED_GENERATOR);
+            goto err;
+        }
+
+        /* look if we can use precomputed multiples of generator */
+
+        pre_comp = group->g2_pre_comp;
+        if (pre_comp && pre_comp->numblocks
+            && G2_ELEM_cmp(group, generator, pre_comp->points[0], ctx) == 0) {
+            blocksize = pre_comp->blocksize;
+
+            /*
+             * determine maximum number of blocks that wNAF splitting may
+             * yield (NB: maximum wNAF length is bit length plus one)
+             */
+            numblocks = (BN_num_bits(scalar) / blocksize) + 1;
+
+            /*
+             * we cannot use more blocks than we have precomputation for
+             */
+            if (numblocks > pre_comp->numblocks)
+                numblocks = pre_comp->numblocks;
+
+            pre_points_per_block = (size_t)1 << (pre_comp->w - 1);
+
+            /* check that pre_comp looks sane */
+            if (pre_comp->num != (pre_comp->numblocks * pre_points_per_block)) {
+                ECerr(EC_F_EC_WNAF_MUL, ERR_R_INTERNAL_ERROR);
```

Comment:
After `bp_err.c` and`macros in `err.h` are in place, how I can trigger `mkerr.pl` to generate the error codes in openssl/bp.h? `make update` does not complain anymore in my local build, but nothing is added to `bp.h`